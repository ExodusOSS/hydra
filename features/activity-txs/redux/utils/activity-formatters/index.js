import { formatSwapActivity } from './format-swap-activity'
import { formatTxActivity } from './format-tx-activity'

export const formattersByType = new Map([
  ['exchange', formatSwapActivity],
  ['tx', formatTxActivity],
])
