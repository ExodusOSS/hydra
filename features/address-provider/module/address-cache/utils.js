import lodash from 'lodash'
import assert from 'minimalistic-assert'
import { set } from '@exodus/basic-utils'

const { get } = lodash

// This allows new assets to use a full key, without regenerating existing data in fusion for legacy ones. Please do NOT add new names here!
const legacyKeyAssetNames = new Set([
  'algorand',
  'ark',
  'avalanchec',
  'bcash',
  'bgold',
  'bitcoin',
  'bitcoinsv',
  'bnbmainnet',
  'bsc',
  'cardano',
  'cosmos',
  'dash',
  'decred',
  'digibyte',
  'dogecoin',
  'elrond',
  'eosio',
  'ethereum',
  'ethereumarbnova',
  'ethereumarbone',
  'ethereumclassic',
  'ethereumgoerli',
  'fantommainnet',
  'filecoin',
  'flare',
  'harmonymainnet',
  'hedera',
  'iconmainnet',
  'lightningnetwork',
  'lisk',
  'litecoin',
  'matic',
  'neo3',
  'monero',
  'nano',
  'nem',
  'neo',
  'ontology',
  'optimism',
  'polkadot',
  'qtumignition',
  'ravencoin',
  'ripple',
  'solana',
  'stellar',
  'terra',
  'tezos',
  'theta',
  'tronmainnet',
  'vechainthor',
  'vertcoin',
  'waves',
  'zcash',
  'zilliqa',
])

export const getCachePath = ({ walletAccountName, baseAssetName, derivationPath }) => {
  assert(typeof walletAccountName === 'string', 'expected string "walletAccountName"')
  assert(typeof baseAssetName === 'string', 'expected string "baseAssetName"')
  assert(typeof derivationPath === 'string', 'expected string "derivationPath"')

  // todo add assetName for all assets when we migrate to a new fusion channel
  let path = derivationPath
  if (!legacyKeyAssetNames.has(baseAssetName)) {
    path = `${baseAssetName}/${path}` // combine these two to match fusion shape
  }

  return [
    // walletAccount is indirectly captured in the BIP path but it is still required because the BIP path is not enough, Trezor's accountIndex always use 1 regardless of the index of a walletAccount.
    walletAccountName,
    // assetName is required because some assets encode the network identifier into the address e.g splitting mainnet vs testnet results in different addresses for the same public key.
    // bipPath is required for assets that may have multiple addresses, or stake addresses etc
    path,
  ]
}

export const diffCaches = (cache1, cache2, fromSync) => {
  let isDifferent = false
  let needsSync = false

  const result = Object.create(null)
  for (const walletAccount in cache2) {
    const subCache = cache2[walletAccount]
    for (const path in subCache) {
      const existing = get(cache1, [walletAccount, path])
      const incoming = subCache[path]
      const newAddress = !existing
      const addressMismatch = !newAddress && existing.address !== incoming.address
      needsSync = needsSync || existing?.synced === false
      const justSynced = !existing?.synced && incoming.synced
      if (fromSync && addressMismatch && !existing?.synced) {
        // We have an address scheduled to sync up, that is mismatching with a
        // sync down, in this case we actually prefer the one scheduled to sync up.
        set(result, [walletAccount, path], existing)
      } else if (newAddress || addressMismatch || justSynced) {
        isDifferent = true
        set(result, [walletAccount, path], incoming)
      }
    }
  }

  return {
    diff: result,
    isDifferent,
    needsSync: isDifferent || needsSync,
  }
}

export const getUnsyncedAddressesForPush = (cache, currentCache) => {
  const out = Object.create(null)
  for (const walletAccount in cache) {
    const subCache = cache[walletAccount]
    for (const path in subCache) {
      const item = subCache[path]
      const currentCacheItem = get(currentCache, [walletAccount, path])
      if (
        currentCacheItem &&
        currentCacheItem.synced &&
        currentCacheItem.address.toString() === item.address.toString()
      )
        continue
      if (!item.synced) set(out, [walletAccount, path], item.address)
    }
  }

  return out
}
