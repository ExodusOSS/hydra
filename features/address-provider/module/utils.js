import { Address, AddressSet } from '@exodus/models'
import lodash from 'lodash'
import assert from 'minimalistic-assert'
import BipPath from 'bip32-path'
import { isNil } from '@exodus/basic-utils'
import KeyIdentifier from '@exodus/key-identifier'

const { groupBy } = lodash

export const resolvePurpose = ({ asset, address }) => {
  // Ideally, We should Unify the creation of Address object to always set the purpose
  if (address.meta.purpose !== undefined) {
    return address.meta.purpose
  }

  // TODO: store address.meta.purpose directly
  const purposeFromIndividualFlags = address.meta.isSegwit
    ? 84
    : address.meta.isNestedSegwit
      ? 49
      : address.meta.isTaproot
        ? 86
        : null

  if (purposeFromIndividualFlags !== null) {
    return purposeFromIndividualFlags
  }

  const addressString = address.toString()
  if (asset.address?.resolvePurpose) {
    return asset.address.resolvePurpose(addressString) || null
  }

  // fallback when resolvePurpose is not implemented (a default implementation could be like below though)
  if (asset.address?.validate) {
    try {
      return asset.address.validate(addressString) ? 44 : null
    } catch (e) {
      // some assets' validate may raise an error when invalid. imho, this should be fixed.
      console.warn(`${asset.name} asset.address.validate function raises an error!!!`, e)
    }
  }

  return null
}

export const getAddressesFromTxLog = ({ asset, txLog, walletAccount }) => {
  const changeAddresses = AddressSet.fromArray(
    [...txLog].map((tx) => tx.data?.changeAddress).filter(Boolean)
  )
    .toArray()
    .filter((address) => address.pathArray[0] === 1) // changeAddress may contain receive address

  const allAddresses = txLog.addresses.union(changeAddresses).sort(AddressSet.PATH_SORTER)

  return [...allAddresses].map((address) => {
    const purpose = resolvePurpose({ asset, address })

    const chainIndex = address.pathArray[0]
    const addressIndex = address.pathArray[1]
    return {
      address: new Address(address.toString(), {
        ...address.meta,
        purpose,
        walletAccount: walletAccount.toString(),
        keyIdentifier: new KeyIdentifier(
          asset.baseAsset.api.getKeyIdentifier({
            purpose,
            accountIndex: walletAccount.index,
            chainIndex,
            addressIndex,
            compatibilityMode: walletAccount.compatibilityMode,
          })
        ),
      }),
      purpose,
      chainIndex,
      addressIndex,
    }
  })
}

export const resolveUnusedAddressIndexesFromAddresses = ({
  addresses,
  purposes,
  highestUnusedIndexes,
}) => {
  const groupByPurpose = groupBy(addresses, 'purpose')
  const knownChains = [0, 1]
  return purposes.map((purpose) => ({
    purpose,
    chain: knownChains.map((thisChainIndex) =>
      (groupByPurpose[purpose] || [])
        .filter(({ chainIndex }) => thisChainIndex === chainIndex)
        .reduce(
          (acc, { addressIndex }) =>
            highestUnusedIndexes
              ? Math.max(addressIndex + 1, acc)
              : addressIndex === acc
                ? acc + 1
                : acc,
          0
        )
    ),
  }))
}

export const inclusiveRange = (end) => {
  return [...Array.from({ length: end }).keys(), end]
}

export const createPath = ({ chainIndex, addressIndex }) => {
  assert(
    !isNil(chainIndex) || isNil(addressIndex),
    'chainIndex must be provided if addressIndex is provided'
  )
  return BipPath.fromString(
    [chainIndex, addressIndex].filter((i) => !isNil(i)).join('/')
  ).toString()
}
