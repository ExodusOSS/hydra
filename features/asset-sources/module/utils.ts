import typeforce from '@exodus/typeforce'
import { getTrezorMeta } from './trezor.js'
import assert from 'minimalistic-assert'
import type { Asset } from '../types.js'
import type { WalletAccount } from '@exodus/models'
import { types } from './types.js'

type GetSupportedPurposesOpts = {
  asset: Asset
  walletAccount: WalletAccount
}

export const getSupportedPurposes = (opts: GetSupportedPurposesOpts) => {
  typeforce(
    {
      asset: types.asset,
      walletAccount: types.walletAccount,
    },
    opts,
    true
  )

  const {
    asset,
    walletAccount: { compatibilityMode, isMultisig },
  } = opts

  const baseAsset = asset.baseAsset
  const apiExists = !!baseAsset.api?.getSupportedPurposes

  assert(
    !isMultisig || apiExists,
    'baseAsset must have api.getSupportedPurposes for a multisig WalletAccount'
  )

  if (apiExists) {
    // we seem to be having two different APIs for getSupportedPurposes, one that expects compatibilityMode separately, and one that expects it on walletAccount
    const supportedPurposes = baseAsset.api?.getSupportedPurposes!({
      ...opts,
      compatibilityMode,
      isMultisig,
    })
    assert(supportedPurposes?.length, 'at least one purpose must be supported')
    return supportedPurposes
  }

  // the given asset should know what's the best value depending on the wallet account type and the compatibility mode.
  // everything below here to be removed

  if (compatibilityMode === 'trezor') {
    const trezorMeta = getTrezorMeta(baseAsset)
    return [
      baseAsset.useBip84 && trezorMeta.isBip84(asset.name) && 84, // segwit
      trezorMeta.isSegwit(asset.name) ? 49 : 44, // nested segwit
    ].filter(Boolean)
  }

  return [
    baseAsset.useBip49 === true && 49, // nested segwit
    baseAsset.useBip84 === true && 84, // segwit
    44, // legacy
    baseAsset.useBip86 === true && 86, // taproot
  ].filter(Boolean)
}

export const getDefaultPurpose = (opts: GetSupportedPurposesOpts) => {
  return getSupportedPurposes(opts)[0]
}
