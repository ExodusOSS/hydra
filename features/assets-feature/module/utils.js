import { CT_STATUS as STATUS } from '@exodus/assets'
import lodash from 'lodash'
import assert from 'minimalistic-assert'

const { get } = lodash

const BLACKLISTED_ERRORS = [/Cannot destructure property 'owner'/u]

export const getAssetFromAssetId = (assets, assetId, baseAssetName) => {
  assert(assetId, 'getAssetFromAssetId(): assetId is required')
  assert(baseAssetName, 'getAssetFromAssetId(): baseAssetName is required')
  // ignore letter-case for assets that support checksummed addresses
  assetId = assetId.toLowerCase()
  return Object.values(assets).find(
    (asset) => assetId === asset.assetId?.toLowerCase() && baseAssetName === asset.baseAsset.name
  )
}

export const isDisabledCustomToken = (asset) =>
  asset.isCustomToken && asset.lifecycleStatus === STATUS.DISABLED

export async function getFetchErrorMessage(error) {
  const defaultExtraMessage = [
    error.message || 'Unknown',
    error.response?.status,
    error.response?.statusText,
  ]
    .filter(Boolean)
    .join(' - ')

  try {
    // Note. I'm not returning responseBody directly because it could be a large error (html) response when the server is down.
    const responseBody = await error.response?.text?.()

    if (!responseBody) return defaultExtraMessage

    const message = get(JSON.parse(responseBody), 'message')
    const isBlacklisted = BLACKLISTED_ERRORS.some((v) => v.test(message))

    if (isBlacklisted) return 'Could not fetch token'

    return message || defaultExtraMessage
  } catch {
    return defaultExtraMessage
  }
}
