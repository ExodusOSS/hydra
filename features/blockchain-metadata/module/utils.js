import lodash from 'lodash'
import { mapValues } from '@exodus/basic-utils'
import { Tx, TxSet } from '@exodus/models'

const { uniqBy } = lodash

export const createTxSet = ({ txs, coinName, assets }) => {
  const values = txs.map((tx) =>
    tx instanceof Tx
      ? tx
      : Tx.fromJSON({
          ...tx,
          // TODO: remove when all code learns how to add coinName explicitly
          ...(tx.coinName ? {} : { coinName }),
          // TODO: remove when all code learns how to add currencies explicitly
          ...(tx.currencies ? {} : { currencies: currenciesForAsset(assets[coinName]) }),
        })
  )
  return TxSet.fromArray(values)
}

// copied from: https://github.com/ExodusMovement/exodus-core/blob/24a359b90bfea80a808e5809e7ac4206b773e525/packages/models/src/tx/utils.js#L43
// After we are all upgraded to @exodus/models v 10, this can be imported from that library.
// For now it is copied to ensure compatibility with older library versions

function assetsForAsset(asset) {
  return Object.fromEntries(
    uniqBy([asset, asset.baseAsset, asset.feeAsset], (asset) => asset.name).map((asset) => [
      asset.name,
      asset,
    ])
  )
}

// Helper for populating the constructor 'currencies' argument
function currenciesForAsset(asset) {
  return mapValues(assetsForAsset(asset), (asset) => asset.currency)
}
