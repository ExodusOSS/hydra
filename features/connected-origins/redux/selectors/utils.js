export const parseOrigin = (origin) =>
  origin
    .replace(/((^\w+:|^)\/\/)?(www\.)?/u, '')
    .replace(/\/$/u, '')
    .split('.')
    .slice(-2)
    .join('.')
