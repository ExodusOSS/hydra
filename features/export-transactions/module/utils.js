export const formatTransactionOutput = ({
  tx,
  isWithdrawal,
  asset,
  coinAmount,
  coinCurrency,
  feeAmount,
  feeCurrency,
  personalNote,
  walletAccount,
  oppositeWalletAccount,
}) => {
  const coinAmountOutput =
    typeof coinAmount?.toDefaultNumber === 'function' ? coinAmount.toDefaultNumber() : ''
  const feeAmountOutput =
    typeof feeAmount?.toDefaultNumber === 'function' ? feeAmount.toDefaultNumber() : ''

  return {
    date: tx.date,
    type: `${isWithdrawal ? 'withdrawal' : 'deposit'}${tx.error ? ' (failed)' : ''}`,
    fromPortfolio: tx.sent ? walletAccount : oppositeWalletAccount,
    toPortfolio: tx.sent ? oppositeWalletAccount : walletAccount,
    outAmount: isWithdrawal ? coinAmountOutput : '',
    outCurrency: isWithdrawal ? coinCurrency : '',
    feeAmount: isWithdrawal || !feeAmount?.isZero ? feeAmountOutput : '',
    feeCurrency: isWithdrawal || !feeAmount?.isZero ? feeCurrency : '',
    toAddress: [tx.to].flat().join(' | '),
    outTxId: isWithdrawal ? tx.txId : '',
    outTxUrl: isWithdrawal ? asset.blockExplorer.txUrl(tx.txId) : '',
    inAmount: isWithdrawal ? '' : coinAmountOutput,
    inCurrency: isWithdrawal ? '' : coinCurrency,
    inTxId: isWithdrawal ? '' : tx.txId,
    inTxUrl: isWithdrawal ? '' : asset.blockExplorer.txUrl(tx.txId),
    personalNote: personalNote?.getMessage({ to: tx.to }),

    tokens: tx.tokens.join(' '),

    // used later, not exported
    sent: tx.sent,
    txId: tx.txId,
  }
}
