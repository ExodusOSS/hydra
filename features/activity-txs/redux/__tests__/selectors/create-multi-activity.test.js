import { OrderSet, Tx } from '@exodus/models'

import { setup } from '../utils'

describe('createMultiActivity', () => {
  test('should return empty array of activities in wallet', () => {
    const { store, selectors } = setup()
    const createSelector = selectors.activityTxs.createMultiActivity({
      createAssetSourceActivitySelector: selectors.activityTxs.createAssetSourceBaseActivity,
    })

    const selector = createSelector({
      assetNames: ['ethereum'],
      walletAccounts: ['exodus_0'],
      foo: 'bar',
    })

    expect(selector(store.getState())).toEqual([])
  })

  test('should map activities, de-dup orders and sort by date', () => {
    const { store, selectors, handleEvent, assets } = setup()

    const BTC_ETH_ORDER = {
      orderId: 'btc-eth-order',
      fromAsset: 'bitcoin',
      toAsset: 'ethereum',
      txIds: [{ txId: 'received-tx' }, { txId: 'sent-tx' }],
      date: '2019-07-22T13:54:28.000Z',
      fromWalletAccount: 'exodus_0',
      toWalletAccount: 'exodus_0',
      toAmount: assets.ethereum.currency.defaultUnit(1),
    }

    const txLogFixtures = {
      exodus_0: {
        ethereum: [
          Tx.fromJSON({
            txId: 'received-tx',
            error: null,
            date: '2019-07-22T13:54:28.000Z',
            confirmations: 1,
            meta: {},
            token: null,
            dropped: false,
            coinAmount: '1 ETH',
            coinName: 'ethereum',
            feeCoinName: 'ethereum',
            feeAmount: '0.000189 ETH',
            currencies: { ethereum: assets.ethereum.currency },
          }),
        ],
        bitcoin: [
          Tx.fromJSON({
            txId: 'sent-tx',
            error: null,
            date: '2019-07-22T13:54:28.000Z',
            confirmations: 1,
            meta: {},
            token: null,
            dropped: false,
            coinAmount: '-1 BTC',
            coinName: 'bitcoin',
            feeCoinName: 'bitcoin',
            feeAmount: '0.000189 BTC',
            currencies: { bitcoin: assets.bitcoin.currency },
          }),
        ],
      },
      exodus_1: {
        ethereum: [
          Tx.fromJSON({
            txId: 'received-tx-2',
            error: null,
            date: '2020-07-22T13:54:28.000Z',
            confirmations: 1,
            meta: {},
            token: null,
            dropped: false,
            coinAmount: '1 ETH',
            coinName: 'ethereum',
            feeCoinName: 'ethereum',
            feeAmount: '0.000189 ETH',
            currencies: { ethereum: assets.ethereum.currency },
          }),
        ],
        bitcoin: [
          Tx.fromJSON({
            txId: 'sent-tx-2',
            error: null,
            date: '2020-07-22T13:54:28.000Z',
            confirmations: 1,
            meta: {},
            token: null,
            dropped: false,
            coinAmount: '-1 BTC',
            coinName: 'bitcoin',
            feeCoinName: 'bitcoin',
            feeAmount: '0.000189 BTC',
            currencies: { bitcoin: assets.bitcoin.currency },
          }),
        ],
      },
    }

    handleEvent('activityTxs', txLogFixtures)
    handleEvent('orders', OrderSet.fromArray([BTC_ETH_ORDER]))

    const createSelector = selectors.activityTxs.createMultiActivity({
      createAssetSourceActivitySelector: selectors.activityTxs.createAssetSourceBaseActivity,
    })

    const selector = createSelector({
      assetNames: ['ethereum', 'bitcoin'],
      walletAccounts: ['exodus_0', 'exodus_1'],
    })

    const result = selector(store.getState())

    expect(result).toMatchSnapshot()
    expect(result.map((i) => i.id)).toEqual([
      'bitcoin.sent-tx-2',
      'ethereum.received-tx-2',
      'swap.ethereum.btc-eth-order',
    ])
    expect(result.map((i) => i.date.toISOString())).toEqual([
      '2020-07-22T13:54:28.000Z',
      '2020-07-22T13:54:28.000Z',
      '2019-07-22T13:54:28.000Z',
    ])
  })

  test('should map activities and dedup items with same id', () => {
    const { store, selectors, handleEvent, assets } = setup()

    // Simulate transfer from exodus_1 to exodus_0
    const txLogFixtures = {
      exodus_0: {
        ethereum: [
          Tx.fromJSON({
            txId: 'some-tx-id',
            error: null,
            date: '2019-07-22T13:54:28.000Z',
            confirmations: 1,
            meta: {},
            token: null,
            dropped: false,
            coinAmount: '1 ETH',
            coinName: 'ethereum',
            feeCoinName: 'ethereum',
            feeAmount: '0.000189 ETH',
            currencies: { ethereum: assets.ethereum.currency },
          }),
        ],
      },
      exodus_1: {
        ethereum: [
          Tx.fromJSON({
            txId: 'some-tx-id',
            error: null,
            date: '2019-07-22T13:54:28.000Z',
            confirmations: 1,
            meta: {},
            token: null,
            dropped: false,
            coinAmount: '-1 ETH',
            coinName: 'ethereum',
            feeCoinName: 'ethereum',
            feeAmount: '0.000189 ETH',
            currencies: { ethereum: assets.ethereum.currency },
          }),
        ],
      },
    }

    handleEvent('activityTxs', txLogFixtures)

    const createSelector = selectors.activityTxs.createMultiActivity({
      createAssetSourceActivitySelector: selectors.activityTxs.createAssetSourceBaseActivity,
    })

    const selector = createSelector({
      assetNames: ['ethereum'],
      walletAccounts: ['exodus_0', 'exodus_1'],
    })

    const result = selector(store.getState())

    expect(result.map((i) => i.id)).toEqual(['ethereum.some-tx-id'])
  })
})
