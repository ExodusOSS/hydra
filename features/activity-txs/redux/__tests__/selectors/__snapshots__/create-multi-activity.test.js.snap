exports[`createMultiActivity should map activities, de-dup orders and sort by date 1`] = `
[
  {
    "assetName": "bitcoin",
    "date": 2020-07-22T13:54:28.000Z,
    "failed": false,
    "id": "bitcoin.sent-tx-2",
    "pending": false,
    "personalNote": undefined,
    "received": false,
    "sent": true,
    "tx": {
      "coinAmount": "-1 BTC",
      "coinName": "bitcoin",
      "confirmations": 1,
      "currencies": {
        "bitcoin": {
          "BTC": 8,
          "bits": 2,
          "satoshis": 0,
        },
      },
      "date": "2020-07-22T13:54:28.000Z",
      "dropped": false,
      "feeAmount": "0.000189 BTC",
      "feeCoinName": "bitcoin",
      "txId": "sent-tx-2",
      "version": 1,
    },
    "txId": "sent-tx-2",
    "type": "tx",
  },
  {
    "assetName": "ethereum",
    "date": 2020-07-22T13:54:28.000Z,
    "failed": false,
    "id": "ethereum.received-tx-2",
    "pending": false,
    "personalNote": undefined,
    "received": true,
    "sent": false,
    "tx": {
      "coinAmount": "1 ETH",
      "coinName": "ethereum",
      "confirmations": 1,
      "currencies": {
        "ethereum": {
          "ETH": 18,
          "Gwei": 9,
          "Kwei": 3,
          "Mwei": 6,
          "finney": 15,
          "szabo": 12,
          "wei": 0,
        },
      },
      "date": "2020-07-22T13:54:28.000Z",
      "dropped": false,
      "feeAmount": "0.000189 ETH",
      "feeCoinName": "ethereum",
      "txId": "received-tx-2",
      "version": 1,
    },
    "txId": "received-tx-2",
    "type": "rx",
  },
  {
    "assetName": "ethereum",
    "date": 2019-07-22T13:54:28.000Z,
    "displayAmount": {
      "type": "NumberUnit",
      "unit": "ETH",
      "unitType": "ETH",
      "value": "1",
    },
    "failed": false,
    "id": "swap.ethereum.btc-eth-order",
    "order": {
      "_version": 1,
      "date": "2019-07-22T13:54:28.000Z",
      "fromAsset": "bitcoin",
      "fromWalletAccount": "exodus_0",
      "orderId": "btc-eth-order",
      "potentialToTxIds": [],
      "status": "",
      "svc": "ss",
      "toAmount": {
        "t": "numberunit",
        "v": {
          "u": {
            "ETH": 18,
            "Gwei": 9,
            "Kwei": 3,
            "Mwei": 6,
            "finney": 15,
            "szabo": 12,
            "wei": 0,
          },
          "v": "1 ETH",
        },
      },
      "toAsset": "ethereum",
      "toWalletAccount": "exodus_0",
      "txIds": [
        {
          "txId": "received-tx",
        },
        {
          "txId": "sent-tx",
        },
      ],
    },
    "pending": true,
    "personalNote": undefined,
    "received": true,
    "sent": false,
    "tx": {
      "coinAmount": "1 ETH",
      "coinName": "ethereum",
      "confirmations": 1,
      "currencies": {
        "ethereum": {
          "ETH": 18,
          "Gwei": 9,
          "Kwei": 3,
          "Mwei": 6,
          "finney": 15,
          "szabo": 12,
          "wei": 0,
        },
      },
      "date": "2019-07-22T13:54:28.000Z",
      "dropped": false,
      "feeAmount": "0.000189 ETH",
      "feeCoinName": "ethereum",
      "txId": "received-tx",
      "version": 1,
    },
    "txId": "received-tx",
    "type": "exchange",
  },
]
`;
